src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs

```diff
@@ -230,44 +231,45 @@ private void SetEntry(CacheEntry entry)
                 }
             }
 
-            StartScanForExpiredItems(utcNow);
+            StartScanForExpiredItemsIfNeeded(utcNow);
         }
 
         /// <inheritdoc />
         public bool TryGetValue(object key, out object result)
         {
             ValidateCacheKey(key);
-
             CheckDisposed();
 
-            result = null;
             DateTimeOffset utcNow = _options.Clock.UtcNow;
-            bool found = false;
 
             if (_entries.TryGetValue(key, out CacheEntry entry))
             {
                 // Check if expired due to expiration tokens, timers, etc. and if so, remove it.
                 // Allow a stale Replaced value to be returned due to concurrent calls to SetExpired during SetEntry.
-                if (entry.CheckExpired(utcNow) && entry.EvictionReason != EvictionReason.Replaced)
+                if (!(entry.CheckExpired(utcNow) && entry.EvictionReason != EvictionReason.Replaced))
                 {
-                    // TODO: For efficiency queue this up for batch removal
-                    RemoveEntry(entry);
-                }
-                else
-                {
-                    found = true;
                     entry.LastAccessed = utcNow;
                     result = entry.Value;
 
                     // When this entry is retrieved in the scope of creating another entry,
                     // that entry needs a copy of these expiration tokens.
                     entry.PropagateOptions(CacheEntryHelper.Current);
+
+                    StartScanForExpiredItemsIfNeeded(utcNow);
+
+                    return true;
+                }
+                else
+                {
+                    // TODO: For efficiency queue this up for batch removal
+                    RemoveEntry(entry);
                 }
             }
 
-            StartScanForExpiredItems(utcNow);
+            StartScanForExpiredItemsIfNeeded(utcNow);
 
-            return found;
+            result = null;
+            return false;
```

Comment:
Did these changes, e.g. avoiding setting a local boolean, really move the needle?

Reply:
I did not compare the performance with my individual changes, just all vs none. My motivation was to get the "hit" case as fast as possible

Reply:
What change is actually contributing the 5-15% improvement?  It just seems like the primary change is tuning inlining and everything else is extraneous; is that not the case?

Reply:
Most probably it's the inlining, as long as setting the `out object result` to `null` does not require a memory barrier

If you believe that this is important I can check this and get back to you, but personally, I am not sure if it's worth it.

Reply:
I'm commenting simply because I want to avoid change for change's sake.  Changes that end up being stylistic decisions, especially in a codebase one's not primarily responsible for, have a way of getting made and then unmade and then made and then unmade, etc.

Reply:
I've checked, and this has removed one additional write barrier:

Before:

![obraz](https://user-images.githubusercontent.com/6011991/99411637-21417000-28f4-11eb-8944-ab9ae7b02bb2.png)

After:

![obraz](https://user-images.githubusercontent.com/6011991/99411721-35856d00-28f4-11eb-96aa-e780e3564651.png)

Reply:
If your goal is to remove write barriers, change the method signature:
```C#
public object TryGetValue(object key, out bool success)
```

Reply:
>  change the method signature:

I wish I could do that. This is a public method and this would be a breaking change ;/

Reply:
Ok.